src/consensus/pool.rs

```diff
@@ -917,4 +932,157 @@ mod tests {
         drop(votor_rx);
         drop(repair_rx);
     }
+
+    #[tokio::test]
+    async fn duplicate_votes() {
+        let (sks, epoch_info) = generate_validators(11);
+        let (votor_tx, votor_rx) = mpsc::channel(1024);
+        let (repair_tx, repair_rx) = mpsc::channel(1024);
+        let mut pool = Pool::new(epoch_info, votor_tx, repair_tx);
+
+        // insert a notar vote from validator 0
+        let vote = Vote::new_notar(0, Hash::default(), &sks[0], 0);
+        assert_eq!(pool.add_vote(vote).await, Ok(()));
+
+        // insert a skip vote from validator 1
+        let vote = Vote::new_skip(0, &sks[1], 1);
+        assert_eq!(pool.add_vote(vote).await, Ok(()));
+
+        // inserting same votes again should fail
+        let vote = Vote::new_notar(0, Hash::default(), &sks[0], 0);
+        assert_eq!(pool.add_vote(vote).await, Err(AddVoteError::Duplicate));
+        let vote = Vote::new_skip(0, &sks[1], 1);
+        assert_eq!(pool.add_vote(vote).await, Err(AddVoteError::Duplicate));
+
+        drop(votor_rx);
+        drop(repair_rx);
+    }
+
+    #[tokio::test]
+    async fn duplicate_certs() {
+        let (sks, epoch_info) = generate_validators(11);
+        let (votor_tx, votor_rx) = mpsc::channel(1024);
+        let (repair_tx, repair_rx) = mpsc::channel(1024);
+        let mut pool = Pool::new(epoch_info.clone(), votor_tx, repair_tx);
+
+        // insert a notar cert for slot 0
+        let mut votes = Vec::new();
+        for v in 0..11 {
+            votes.push(Vote::new_notar(0, Hash::default(), &sks[v as usize], v));
+        }
+        let notar_cert = NotarCert::try_new(&votes, &epoch_info.validators).unwrap();
+        assert_eq!(pool.add_cert(Cert::Notar(notar_cert.clone())).await, Ok(()));
+
+        // insert a skip cert for slot 1
+        let mut votes = Vec::new();
+        for v in 0..11 {
+            votes.push(Vote::new_skip(0, &sks[v as usize], v));
+        }
+        let skip_cert = SkipCert::try_new(&votes, &epoch_info.validators).unwrap();
+        assert_eq!(pool.add_cert(Cert::Skip(skip_cert.clone())).await, Ok(()));
+
+        // inserting same certs again should fail
+        assert_eq!(
+            pool.add_cert(Cert::Notar(notar_cert)).await,
+            Err(AddCertError::Duplicate)
+        );
+        assert_eq!(
+            pool.add_cert(Cert::Skip(skip_cert)).await,
+            Err(AddCertError::Duplicate)
+        );
+
+        drop(votor_rx);
+        drop(repair_rx);
```

Comment:
for my knowledge really, why do we need to drop these?  Won't they get dropped when they go out of scope below?